Merge from Base Branch Before E2E Runs on GitHub Actions

U8NWXD opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #[n/a].
2. This PR does the following: Before running E2E tests on actions, merge from the base branch of a PR to incorporate any fixes since the last time the developer merged from develop.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

See the CI runs on this PR ([example build](https://github.com/oppia/oppia/pull/12829/checks?check_run_id=2605793156)) for proof that these changes work for PRs across forks. For proof that these changes work for PRs within a single repository, see U8NWXD/oppia#14 ([example build](https://github.com/U8NWXD/oppia/pull/14/checks?check_run_id=2605800807)).

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

U8NWXD committed:
Create action for merging from base branch on PRs

U8NWXD committed:
Add merging from base action to all e2e tests on actions

U8NWXD committed:
Replace workflow if with bash if

U8NWXD committed:
Remove workflow if

U8NWXD committed:
Only fetch base branch

U8NWXD committed:
Remove git show since merge already prints information

U8NWXD committed:
Remove checkout since we want to run tests on feature branch

U8NWXD committed:
Add debugging git log

U8NWXD committed:
Checkout head branch before merging in base

U8NWXD committed:
Get head branch from environment variable

U8NWXD committed:
Reference environment variable directly instead of through context

U8NWXD committed:
Add debugging dump of github context

U8NWXD committed:
Try getting head repo from github context

U8NWXD committed:
Fetch source branch

U8NWXD committed:
Do not limit merging to fast-forwarding

U8NWXD committed:
Configure git identity

U8NWXD committed:
Remove debugging step

U8NWXD commented:
@DubeySandeep PTAL

DubeySandeep commented on .github/workflows/e2e_learner_flow_skill_editor_and_embedding.yml:
```diff
@@ -23,6 +23,8 @@ jobs:
           architecture: 'x64'
       - name: Setup python by installing wheel
         run: pip install wheel==0.35.0
+      - name: Merge from Base Branch if Running on a PR
+        uses: ./.github/actions/merge-base-branch
```
Do we need the name property here? will it use the name defined in the action itself?

DubeySandeep replied:
Any specific reason for doing it after installing the wheel?

U8NWXD replied:
Yes, I think it will use the action name if we leave out `name`. It doesn't matter whether we do this before or after installing wheel. I put the step here because this is right before we start referencing Oppia code

U8NWXD replied:
done

DubeySandeep commented on .github/actions/merge-base-branch/action.yml:
```diff
@@ -0,0 +1,26 @@
+        echo "Base Branch: ${{ github.base_ref }}"
+        echo "Head Repository: ${{ github.event.pull_request.head.repo.full_name }}"
+        echo "Head Branch: $GITHUB_HEAD_REF"
+        git config --global user.email "admin@oppia.org"
+        git config --global user.name "Oppia Merge Base Branch GitHub Action"
```
Do we need this step, if yes should we use dummy email/username?

U8NWXD replied:
Yes we need this. Without it, `git` will refuse to make the commit. This commit exists only for the duration of the CI check since we don't push it anywhere. Therefore, I don't think it really matters what the email and username are. I made the user name descriptive so that anyone debugging (e.g. by adding a step to execute `git log`) would know where it came from

DubeySandeep replied:
Leaving it to admin to decide, @seanlip Do you think it's fine to use this email?

seanlip replied:
No, I don't think so. I suggest using a made-up email address that explains what it is for, e.g. tempaccount.for@merge.check or similar, since these commits are not being made by the Oppia admin.

Also, it would be better not to need these statements at all if possible. What error message exactly do you get when this is not set?

U8NWXD replied:
This is required because merging creates a merge commit, and creating a commit requires a name and email. If we don't do this configuration, we'll get an error like this:

```
*** Please tell me who you are.

Run

git config --global user.email "you@example.com"

git config --global user.name "Your Name"

to set your account's default identity.

Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got '...')
```

U8NWXD replied:
I've changed the email to use `example.com`

DubeySandeep reviewed (changes requested):
LGTM, I've left a few questions, PTAL!

U8NWXD committed:
Checkout base (e.g. develop) and merge in source

We want to run tests on the commit formed by merging the feature branch
into develop, not the other way around, because that commit matches the
one that will be created when the PR is merged.

U8NWXD committed:
Remove name when calling merge action

U8NWXD commented:
@DubeySandeep PTAL. I've addressed your comments

DubeySandeep commented:
@seanlip I've left a comment for you over [here](https://github.com/oppia/oppia/pull/12829#discussion_r639215646), PTAL!

DubeySandeep commented on .github/actions/merge-base-branch/action.yml:
```diff
@@ -0,0 +1,26 @@
+          echo "Skipping merge because not triggered by a PR event."
+          exit 0
+        fi
+        echo "Base Repository: ${{ github.repository }}"
```
Just to confirm we are expecting Base repo to be `oppia/oppia`, right?

U8NWXD replied:
It usually will be, but sometimes contributors (like me!) open PRs against their own forks for debugging since those tests bypass the oppia/oppia queue

DubeySandeep commented on .github/actions/merge-base-branch/action.yml:
```diff
@@ -0,0 +1,26 @@
+        git fetch source $GITHUB_HEAD_REF
+        git fetch base ${{ github.base_ref }}
+        git checkout base/${{ github.base_ref }}
+        git merge source/$GITHUB_HEAD_REF
```
In the container, we have the contributor's feature branch, right? (not sure) and we need to merge with `oppia/oppia develop`, right? OR is it the opposite? If the latter one, then how it is working as of now? [not sure!]

U8NWXD replied:
In the container, we have a merge commit GitHub creates by merging the feature branch into `develop`. However, this merge commit doesn't get updated when we push commits to `develop`. With this PR, we stop using GitHub's merge commit and instead do the merge ourselves to ensure we get the latest `develop`.

DubeySandeep commented on .github/actions/merge-base-branch/action.yml:
```diff
@@ -0,0 +1,26 @@
+        git fetch source $GITHUB_HEAD_REF
+        git fetch base ${{ github.base_ref }}
+        git checkout base/${{ github.base_ref }}
+        git merge source/$GITHUB_HEAD_REF
```
The Name of the action is `merge from the base branch` but over here we are merging with the source, is that expected? or I'm missing something?

U8NWXD replied:
We are merging the source branch into base. I'll update the action name (it used to merge the base into source)

U8NWXD replied:
fixed

DubeySandeep reviewed (changes requested):
LGTM, left a few questions! (The changes LGTM!)

U8NWXD committed:
Rename action now that we aren't merging from base

U8NWXD committed:
Merge branch 'develop' into merge-base-ci

U8NWXD committed:
Change commit email

U8NWXD commented:
@DubeySandeep PTAL. I've addressed your comments

U8NWXD committed:
Merge branch 'develop' into merge-base-ci
